src/main/java/com/intellij/rt/debugger/agent/CaptureStorage.java

```diff
@@ -23,98 +23,196 @@ protected Deque<CapturedStack> initialValue() {
     }
   };
 
+  private static final ThreadLocal<Boolean> THROWABLE_CAPTURE_DISABLED = new ThreadLocal<Boolean>() {
+    @Override
+    protected Boolean initialValue() {
+      return false;
+    }
+  };
+
   @SuppressWarnings("StaticNonFinalField")
   public static boolean DEBUG; // set from debugger
   private static boolean ENABLED = true;
 
+  private static final StackTraceElement ASYNC_STACK_ELEMENT =
+          new StackTraceElement("--- Async", "Stack.Trace --- ", "captured by debugger agent", -1);
+
   //// METHODS CALLED FROM THE USER PROCESS
 
   @SuppressWarnings("unused")
-  public static void capture(Object key) {
-    if (!ENABLED) {
+  public static void capture(final Object key) {
+    withoutThrowableCapture(new Runnable() {
+        @Override
+        public void run() {
+            captureInt(new Throwable(), key);
```

Comment:
isn't is better to move new Throwable() to the very beginning of the capture method before withoutThrowableCapture call?

Reply:
No, it's not possible because of recursion problems.